### ISSUE TITLE
[BUG] Time series predicrtor improper cache is used when prediction_length is changed

### ISSUE THREAD

**Issue by WenjianDong** [autogluon/autogluon#3986]
**Bug Report Checklist**

- [ ] I provided code that demonstrates a minimal reproducible example.  
- [ ] I confirmed bug exists on the latest mainline of AutoGluon via source install.  
- [x] I confirmed bug exists on the latest stable version of AutoGluon.  

**Describe the bug**
The `predict` function of `autogluon.timeseries.TimeSeriesPredictor` ([source code](https://github.com/autogluon/autogluon/blob/2a5b805d32d9468ab6d50f269b6cd504beaf6414/timeseries/src/autogluon/timeseries/predictor.py#L741)), it first checks the cache ([source code](https://github.com/autogluon/autogluon/blob/2a5b805d32d9468ab6d50f269b6cd504beaf6414/timeseries/src/autogluon/timeseries/trainer/abstract_trainer.py#L935)). However, to determine whether a proper cache exists, the current latest mainline branch only checks train data, known covariates, and static features([source code](https://github.com/autogluon/autogluon/blob/2a5b805d32d9468ab6d50f269b6cd504beaf6414/timeseries/src/autogluon/timeseries/trainer/abstract_trainer.py#L1003)). The `prediction_length` is not checked. This means, if we change the prediction length but keep train data, known covariates, and static features unchanged, it will not return the prediction of the correct length. 

Currently we can overcome this bug by setting `cache_predictions` to False, or manually deleting the cache file each time when merely changing the prediction_length. However, this will give up the benefit of caching, i.e. saving time for repetitive computation. 

I imagine a possible fix can be adding prediction length to `cache_predictions` at [here](https://github.com/autogluon/autogluon/blob/2a5b805d32d9468ab6d50f269b6cd504beaf6414/timeseries/src/autogluon/timeseries/trainer/abstract_trainer.py#L1003). 

**Installed Versions**
v1.0.0

**Comment by aaronschiff** [comment]
I believe this affects other options of `TimeSeriesPredictor`, not just `prediction_length`. For example changing `eval_metric` also does not seem to invalidate the cache. Also changing options for individual models via the `hyperparameters` argument to `TimeSeriesPredictor.fit()` may not invalidate the cache (e.g. changing the `max_p` argument for an `AutoARIMAModel`). 

For now my approach is just to set `cache_predictions=False` when creating a `TimeSeriesPredictor`. Or you can set `use_cache=False` when calling `TimeSeriesPredictor.predict()`.

**Comment by shchur** [comment]
Hi @WenjianDong, thanks a lot for the detailed issue description!

Changing the parameters of `TimeSeriesPredictor` such as `prediction_length` or `eval_metric` after the predictor is created & `fit()` is called will have no effect on the behavior of the predictor. For example
```python
predictor = TimeSeriesPredictor(prediction_length=10).fit(data)
predictor.prediction_length = 1
forecast = predictor.predict(data)  # still contains predictions 10 steps into the future, not 1
```
If you want to change the `prediction_length` after fitting the predictor, you either need to subsample the data with `predictions.groupby("item_id").head(new_prediction_length)` (only works for smaller length), or create & train a new predictor with the new configuration.

Therefore, I guess the real problem is not in the caching behavior, but rather in
 1. the fact that we allow changing predictor parameters after fitting and
 2. lack of documentation for this behavior

As a potential solution, we need to expose the immutable properties of the `TimeSeriesPredictor` as `@property`, similar to how [`TabularPredictor` does it](https://github.com/autogluon/autogluon/blob/master/tabular/src/autogluon/tabular/predictor/predictor.py#L296).

**Comment by aaronschiff** [comment]
@shchur In my opinion, it's normal when setting up a forecasting process to experiment with parameters, and a natural way to do this is to make manual adjustments and re-run an existing `TimeSeriesPredictor` rather than create a new one. Perhaps you could consider to make caching the predictions optional rather than the default? Then users could actively enable the cache when they really need it. Either way, I agree that the documentation could be improved. Thanks for your help!

**Comment by shchur** [comment]
@aaronschiff the problem is that the `prediction_length` parameter is "baked in" into most models. For example, if we train PatchTST with some `prediction_length`, then its output projection layer will have shape `[hidden_dim, prediction_length]`. Even if we change `prediction_length` of the `TimeSeriesPredictor` object, that will have no effect on the PatchTST model, so we'll need to train a new model from scratch to modify the `prediction_length`. Similar thing happens for `eval_metric`, where the models might be trained differently, depending on which `eval_metric` was provided at creation time. 

Therefore, we currently do not support changing these parameters after `fit()` is called since we can't really propagate these changes to the trained models.

**Comment by shchur** [comment]
I now understand that the problem was likely an instance of #4183. This has been fixed in the v1.1.1 release, so I'm closing the issue for now.
